Mathlib/Order/Filter/CountableSeparatingOn.lean

```diff
@@ -0,0 +1,219 @@
+closed sets), all theorems in this file take a predicate `p : Set α → Prop` as an argument and prove
+existence of a countable separating family satisfying this predicate by searching for a
+`HasCountableSeparatingOn` typeclass instance.
+
+## Main definitions
+
+- `HasCountableSeparatingOn α p t`: a typeclass saying that there exists a countable set family
+  `S : Set (Set α)` such that all `s ∈ S` satisfy the predicate `p` and any two distinct points
+  `x y ∈ t`, `x ≠ y`, can be separated by a set `s ∈ S`. For technical reasons, we formulate the
+  latter property as "for all `x y ∈ t`, if `x ∈ s ↔ y ∈ s` for all `s ∈ S`, then `x = y`".
+
+This typeclass is used in all lemmas in this file to avoid repeating them for open sets, closed
+sets, and measurable sets.
+
+### Main results
+
+#### Filters supported on a (sub)singleton
+
+Let `l : Filter α` be a filter with countable intersections property. Let `p : Set α → Prop` be a
+property such that there exists a countable family of sets satisfying `p` and separating points of
+`α`. Then there `l` is supported on a subsingleton: there exists a subsingleton `t` such that
+`t ∈ l`.
+
+We formalize various versions of this theorem in
+`Filter.exists_subset_subsingleton_mem_of_forall_separating`,
+`Filter.exists_mem_singleton_mem_of_mem_of_nonempty_of_forall_separating`,
+`Filter.exists_singleton_mem_of_mem_of_forall_separating`,
+`Filter.exists_subsingleton_mem_of_forall_separating`, and
+`Filter.exists_singleton_mem_of_forall_separating`.
+
+#### Eventually constant functions
+
+Consider a function `f : α → β`, a filter `l` with countable intersections property, and a countable
+separating family of sets of `β̀`. Suppose that for every `U` from the family, either
+`∀ᶠ x in l, f x ∈ U` or `∀ᶠ x in l, f x ∉ U`. Then `f`` is eventually constant along `l`.
+
+We formalize three versions of this theorem in
+`Filter.exists_mem_eventuallyEq_const_of_eventually_mem_of_forall_separating`,
+`Filter.exists_eventuallyEq_const_of_eventually_mem_of_forall_separating`, and
+`Filer.exists_eventuallyEq_const_of_forall_separating`.
+
+#### Eventually equal functions
+
+Two functions are equal along a filter with countable intersections property if the preimages of all
+sets from a countable separating family of sets are equal along the filter.
+
+We formalize several versions of this theorem in
+`Filter.of_eventually_mem_of_forall_separating_mem_iff`, `Filter.of_forall_separating_mem_iff`,
+`Filter.of_eventually_mem_of_forall_separating_preimage`, and
+`Filter.of_forall_separating_preimage`.
+
+## Keywords
+
+filter, countable
+-/
+
+open Function Set Filter
+
+/-- We say that a type `α` has a *countable separating family of sets* satisfying a predicate
+`p : Set α → Prop` on a set `t` if there exists a countable family of sets `S : Set (Set α)` such
+that all sets `s ∈ S` satisfy `p` and any two distinct points `x y ∈ t`, `x ≠ y`, can be separated
+by `s ∈ S`.
```

Comment:
Can you be more specific about what this means to be separated here? (it's not: "there exists `s` such that `x` belongs to `s` and `y` doesn't", its "there exists `s` such that `x` belongs to `s` and `y` doesn't or conversely")